test/extensions/resource_monitors/downstream_connections/downstream_connections_monitor_test.cc

```diff
@@ -0,0 +1,97 @@
+#include "envoy/extensions/resource_monitors/downstream_connections/v3/downstream_connections.pb.h"
+
+#include "source/extensions/resource_monitors/downstream_connections/downstream_connections_monitor.h"
+
+#include "absl/types/optional.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace DownstreamConnections {
+
+class ActiveDownstreamConnectionsMonitorTest : public testing::Test {
+public:
+  void initialize(const envoy::extensions::resource_monitors::downstream_connections::v3::
+                      DownstreamConnectionsConfig& config) {
+    monitor_ = std::make_unique<ActiveDownstreamConnectionsResourceMonitor>(config);
+  }
+
+  Thread::ThreadSynchronizer& synchronizer() { return monitor_->synchronizer_; }
+
+  std::unique_ptr<ActiveDownstreamConnectionsResourceMonitor> monitor_;
+};
+
+TEST_F(ActiveDownstreamConnectionsMonitorTest, CannotAllocateDeallocateResourceWithDefaultConfig) {
+  envoy::extensions::resource_monitors::downstream_connections::v3::DownstreamConnectionsConfig
+      config;
+  initialize(config);
+  EXPECT_FALSE(monitor_->tryAllocateResource(1));
+  EXPECT_EQ(0, monitor_->currentResourceUsage());
+}
+
+TEST_F(ActiveDownstreamConnectionsMonitorTest, ComputesCorrectUsage) {
+  envoy::extensions::resource_monitors::downstream_connections::v3::DownstreamConnectionsConfig
+      config;
+  config.set_max_active_downstream_connections(10);
+  initialize(config);
+  EXPECT_EQ(0, monitor_->currentResourceUsage());
+  EXPECT_EQ(10, monitor_->maxResourceUsage());
+  EXPECT_TRUE(monitor_->tryAllocateResource(3));
+  EXPECT_EQ(3, monitor_->currentResourceUsage());
+  EXPECT_TRUE(monitor_->tryDeallocateResource(2));
+  EXPECT_EQ(1, monitor_->currentResourceUsage());
+}
+
+TEST_F(ActiveDownstreamConnectionsMonitorTest, FailsToAllocateDeallocateWhenMinMaxHit) {
+  envoy::extensions::resource_monitors::downstream_connections::v3::DownstreamConnectionsConfig
+      config;
+  config.set_max_active_downstream_connections(1);
+  initialize(config);
+  EXPECT_EQ(0, monitor_->currentResourceUsage());
+  EXPECT_EQ(1, monitor_->maxResourceUsage());
+  EXPECT_TRUE(monitor_->tryAllocateResource(1));
+  EXPECT_FALSE(monitor_->tryAllocateResource(1));
+  EXPECT_TRUE(monitor_->tryDeallocateResource(1));
+  EXPECT_EQ(0, monitor_->currentResourceUsage());
```

Comment:
seems like we test alloc at max, but not dealloc while at min? Can we add that here -- seems like it'd fit what the test case name describes